### ISSUE TITLE
Create a Vizro app where the pydantic configuration is translated to yaml

### ISSUE THREAD

**Issue by huong-li-nguyen** [mckinsey/vizro#726]
**Requirements:**
- Simple Vizro app
- Text input field (accepts yaml or pydantic)
- Text output field (Returns yaml or pydantic)
- Action that translated from one to the other
- Host it on huggingface or similar
- Add your app to the community demos

**Comment by antonymilne** [comment]
LOVE this idea! FYI @maxschulz-COL we discussed similar before.

Note the to YAML version won't be possible yet until we implement the full `to_dict` or `to_yaml` functionality, which we haven't prioritised yet and maybe best to wait for our move to pydantic v2.

**Comment by huong-li-nguyen** [comment]
I'd love to have this because I think it would also speed up our dev workflow 😄. Creating the YAML examples for our docs takes a lot of time and isn't much fun, so automating it would be awesome.

The standard Pydantic to YAML converters are okay and give a good starting point, but they don't handle all the Vizro-specific syntax like `__target__` and stuff.

**Comment by maxschulz-COL** [comment]
> No, not yet. It's something that would still be nice to have.

Do you mean because of the MM? Because one way its possible now right? yaml -> pydantic

**Comment by antonymilne** [comment]
Yes, that way is possible. It's the opposite direction that doesn't exist. This isn't a limitation of the model manager, it's just because if we do `dashboard.model_dump(_json)` then it doesn't yet know how to handle `CapturedCallable` and probably some other things.

**Comment by huong-li-nguyen** [comment]
Yes, I was mostly interested in the opposite direction — converting **Pydantic → YAML**. This is mainly out of personal interest, to be honest. I’ve been writing a lot of documentation, tutorials, and dashboard examples lately, and creating and testing the YAML examples is always the most time-consuming (and least enjoyable 😄) part of the process.  

I agree it’s not a high priority and more of a “nice to have”.

**Comment by maxschulz-COL** [comment]
> Yes, that way is possible. It's the opposite direction that doesn't exist. This isn't a limitation of the model manager, it's just because if we do `dashboard.model_dump(_json)` then it doesn't yet know how to handle `CapturedCallable` and probably some other things.

Yes that is also a limitation, but I was thinking if like Li describes at the top, we have a simple Vizro app that does this, once we instantiate the model and thus the MM in order to do `dashboard.model_dump(_json)`, then we might get conflicts no?

**Comment by antonymilne** [comment]
I also got bored writing docs yaml examples and so cobbled together something very quickly that does this. I used it for all examples in #1379 and it worked perfectly there.

[Here's an example](https://py.cafe/snippet/vizro/v1?#c=H4sIADMI6WgEA71Ze2_bthb_KoQ74MqFrfmRlwN4u223db3Y3Yo1uP0jDjRaom02EqVJVGwlyHe_v8OHLbtOtmFD2QYRycPz-J0HH3noxHkiOpedF_9Im6kX7J2qRKlZk9clS3i1mue8TBiJYVKxlSjFJcj-kTZTMityCLuT92UeZpCRVoxX7C47mCrSXKdNKDZFKSpDUmxmalHmmSVgjvp_1GlPhJrPU1H5eTIo4stoWcpkj8zKDnVT7IhjXui6FHt0C7nE0JbktpBRDHwO1OWxlrmypvCZmikti4pNoXSYcM1D6gZdmpipfzsxwayzLHmxmnUwkYgFm_MyygRXAS2JFiXPRI9teqzpwgEMLVnAlCXY7gjCZZnXxbwJNt0Qc8FDc8lmHeIy6zx2Q2gudCTBfkPiiQkMsopBXGA5QsqU5EybHkv5HLhMH2Yd6DzrELdXd6LkS8GuZMGCr7rEeMsqrAsoIyJd8lhUwVomejUdhGeOohSwVBGhtX0hy0pHBXGbwunhe3wFllRLnYrprPMdbJt1enYwzoGyEkpX02s7Qg0LXy3fwqNuqW8QA1ynbZ8HhHzXcfNtkedalJBF_16-JInsA8I_FpcvX7LXZcOV7rH3IXsbMq4S9iGTetVj_2XBcDI5dab59h6WaxnzlBk-rxRPm0pWl-wNrwT7oOtEUvgo9rqupKJg_qC5lhXWVOE-qx_zTKzzPOmxdz9dsl9lvKJM_C5k78o1GLyv56msVlItQ6t6y6y2iYDnda11rgItNhpmfr8xoeqQZS5Yp3ec8gszEcVT0PU8bvAbRs5QF-JcJc-760OdZbxstsqkHIVET0FsHEQ-mF5fD3ps2GN983PTY9ejHnP_n-3deJWejIMfjMuPx4HP1YNZarsEmlKAOCntdsfTWkRxntaZgpk61zyN5jJNt5a2G5IoWtQqBqXNvSM0luMiLzNOTvnqwQxchqPF41F6j7BPv9fHhLeC0WNF7UsiQ2XiCNmXgAQVqZp19umeRuSKz6sDi7FZ7EWUb6B-kyvNkbDlwRLfvC5XFBnGObY4tmS22_EIPmyQ-5b2hMA5CmV6hVKRY6PITCmjUg3MW9G4TdvDRmlsPlrtGO1ftpVQ_4K2Unz9LSPbNC46_BDqmy5z7HfX0OYHmWJbCLaRnXCUtW6PFD2c0jITMJ9VIhWxzkuqd29WIr5FgdZUSY8tquQ9FnVvXHXVK1n-QXH1m8kW6aerq6-cf6JmWtR3Q9TAZ9bBiWAryTevykc6CLIkZ2vBlkKzLEeXvPTtZ0ucP7fHGevLBkyML61bHbi7le3v55S8k3kqPy8nXs93FdMrwbirEEkp74Ri84ZxthBrBvRSKcqntUYUWgnPqN0Dtimcvu3O8830qqxF24b293P2rATXGScB-5PeoI8rnAOYOeYxiiAmKwt-UeYLaU67zxmTCFVJ3UROTCuxnO5koI3MHrvHtzOWUtGVb363pBE6TES81jmGvkKJPu6-zzNrRwMY3nPaWigtdsPUNC8RVUTtHBxuSOa2ZwC3I84ST4Awo08v17d2Zv7KE5m_0yI7LP_U8sKehiDZIUIY2PQGMsKJqbL8VjgNLFzIOLNdAQ630LvszYorxN6mzzdwloQDcJeiw5yuLmeH21VLb_9NtpgSsbuOmQrxne86M6h2QPHdoZoK0vbIBn22Fcaj4zU0JdxXFxwMHV8YYekUv5NLnFJzRfj9vO218ANJ1AYZRK9xo9gRUJMEejsEfLMLfpLqNjA3Dujkj6ekr3GHHQDMEjY5gjkO1fuB59sRjvADIN_juTuuwo8egLYQ2MDMus-loHCb32huGyEKfJq77d9u5HB3qWx4lm47n6ocM5DCXuGMAZ9WbC41W5dSa1vaoK9--_6KzWuNAADiTKNO5-UtinWahiGuGXS_xDVf8hSxG-TzT_5OKRcoKAhRzVVsJszWZa7F4Rt7UU3e8BSQpsKvofbC3-04S2Ss0aObOnYbEyc7OkeF-2RkczyiSyXkhOZ0SMRhFCkUhSjqsZcvzQwI64x2rkfLqOQSd6krKPV9WeZlgNpDHfZAehprHpnKyXhrISkL_5kUguEwDy8ckSGmAOqxHNsD9s6tRfsoWBoyq2vuf60pv9BN7wHyA-VhDxbHdQl1pYIXgBy7FY25AJrY3dIDrTrVSOwHZyQ1HIXZLTzgyUOTQEFbDjW79vr2Bsvbxt3tLAsBdQBWD49dF6zUXrCrlVAIdrMIKUDLZh1mHjz8UkLDOZNxbSMuFXcCIWmYoIFiuxaq-pXH1fSUpK2n3A3ulrhQsavssEiPeYYOWc94xky3VKFQtS5BuaKw6q8pnAhrIsVZllcV4g37Ktcx3a9hrlrqVdVCzil3bZFzsQS4u4YPfREQ97I4jDBnIB5V2vA4dkRLCqYCBxZW4RWBlxX7GsGs-iat7CepibelPkqVKQQfBVtz8g6cgzNOnRUiYRBYIZso98UmTms4soayGocBkSYoGqgOUtmJ30j_30KKBpwkklxU6l-2dtDyOc5PigleNRDT0AjOfSYMSPgajyGsaBIoIGMWZLwBfZFXlUTWMXrkWCIVILJvgmYrU6_yeolnFJOoKCvfdkO8utB58lMN8zKuamRLg9MlLiJ4U8KpCoLIyzBSGdGa3xIf_EB9yis6z7F8Ycbs1sdw5LWhYUPcZCFNm12SKiFgHSISt3urfQqMCMXAARcZ4OyRzq4YPbUC8zZgQdCODSPGVokRaEj7dwsydl7jWgCNgKbRzbxgkobwHA2Cha1hVBpZBdDShDxi15klh3UZ22wl0gXKJNiUYAPvKSES8gTISXaNOMOmgRCl7SQk1en1iTpRpUsovxsPrEFQXiw4PqZeJQKDmP2CUKPCBmsQGFtYcCwFLKz_jbGafhNPBvakSGuYtjc3HBrt2sh4YdZ-Y-8Obm_pDq7Y4GMiCCAlxvvVczq1hZNsmwIcPwg3oIeIN_tnIlE18Ab9hIktNixAtQ23KdGnqZ4hCHmSRKXfGUXZhY2aDu5zuK0xmYBtGlFOihQoUbVBisBwjkdIacETxLoo2H8-_PLz18YOmEjP9bNOZZ4tTQqUtcZb8l4Q2GS0qhYilgskrBbxSsnfa-zvM1VAfx0YVQnewARBmvOkMp8UGrjLSuyiSk9HtJPMlAnYoBuSuxK8VTto8Nhcq6Db6XVK8XsN8MwGjr9UmBfyb_AoPAxPTiGyIXEgozTx0-gWDQqEwkDR5AkKZ_9uEI7OwiGm7K23c_nQcTWuczmCGLzi_pqvadixKtHrdVDB06QUILrezuDgiITGpHmg7lwOTwe9TibVR9sd296PQi5XkENdmWDZQqbiNbgiJrdPJE9IINL-3NKCpIAXMNp5vHnsbWl2WjhBw-FJeD4cTYYXZ2eD0cX5YHKcORbuePKiCIums893N91GP9Qbva-ChcjLvzgNSerFcDyeGD2eFL-v92gcTi4Go9HoZHI2GUzOz07-DJoOnT9C0oO4p_dnKgzPz8LBcHJ-fn46GA8mgxFUWLk5mPJn9EFCYpinf6SQpyON6N9jz3gbsX19g-DM41vqEhdgDvoXbIWkuGTDCzE5Ox-MkuFcnOAvB-ICfz2Ynw3EWAwu-GgwU1yhBGCrS_okqprSX05CGscte5Fis5tOR-E4HM7UPOXxLXqnyCIQ4A8BCtfR6XQYTqgf83iFs9KcOAzHZgQ7EfIdSwBNOApHGMK1BiD2VY73VirnWD8Ge5pKJbG_APdzbHhQn6ZGxIg6fb7sm7em6djpY0bnyEBUCF70W29N0HgQnjgKbOoayB7Mj4gD7KsgEtxIAdPrkxlSLYnGSJGJ4kQCnexlCCb28XzA6ZWa1D0nDbEnoPIuBZ5IHHsMfpLqEx9Z_uOZwqXvti4qvjCQDkMUoqwpmj7eNOiBhB4AgKXBVvFyveJ4uQDhqRkQle7jMKRimROVcZFCuWyohxoFawscOfhOOuUiVVzjD2NukXKNk2KGbQVkJ-GJITN_5ZxOz2Audd0OQmyMZX4A-OE1B6NDeIiGTXCAz9jxcSVzpXVBIq1EU1D7gErUWqa0HLES4qSmzRJ9jyEKDkNLRZkUgb50oqH9wTprbEDBebVsjGsAJxmMQKoLndMLz_RsAqCwrJIbmh8agBDRoN8DGKIMM31v_eel12UKx45JmtHcbAkGOrNjrEV5ey9qyHbOXCPk9E4SDt2wGuFqQKwq3FM6j_8HkalBkykfAAA) of the [tips tutorial dashboard](https://py.cafe/vizro-official/vizro-tips-analysis-tutorial) being converted to YAML, which is a bit more complicated. Here it also works well apart from navigation which uses the random ids for pages rather than page names. Look at the console output on pycafe for the yaml version.

Note you need to manually replace `data_frame: <random-id>` with `data_frame: tips` or similar. The custom `bar_mean` function is possible with `_target_: __main__.bar_mean`.

My feeling now is that we don't want a dashboard that does pydantic -> yaml conversion but instead just have some function `to_json`/`to_yaml` like we have `to_python`. My hacky code isn't the right way to do it but I don't think it would be hard to do properly.

[Another more up to date example](https://py.cafe/snippet/vizro/v1?#c=H4sIAApq6WgEA61Y23Ljthl-FYx8sVJGog4-yPaM2m6dTbK92N1J3ObC8jAQCYlcUwADgJK5Hr9EX6DTN-wj9PsBnqTYSWYSeDwSgf_4_SdQT71IxaJ33Tv5U9ZSnrD30ghtWakKzWJukpXiOmakhqWSJUKLa5D9KWsp022uoGyXftEqyDNlszIQj7kWxjBuWP54RLKFGZk72m2PjnhkUyX9GV_KtVbbA6bAlrkwrOKJeG4LLbp09dG_6KF7EFi-ylpWAiXkm3Cj03gpl3LD820qY6HZAgYHMbc8aPb6g-DnQuiyv0SkpE2lkJYtFuzNu0KrXLxhXMasFBzMCzabTObL3oCE_q0yEHwbzfPEbcdizUzErRU63Kc2CZN0k2T4t33SGq4134oha3bDSBXS6nLxQUkxuGZsKRlWtRumphUgYljfCrkjcx3VsndPpjWEtcxa2DrdeL8rw_p-m1YrbthuPi7gUZx_EhoeLnudkxInWboW7x4P9036ReAoV7_cDrf8cXEx6eyqnEepLReT4LSzG6lM6cXLfnfJuBUbBQqlETqzeCIUwLnsXbO7b3hmAO6tLsT987B23wWLvqTrFzBKDZPKMoL_umKgBcyCIgc8IrSaR8L0jchEZGHjdMi2XD8ITcozpAsCHVuE_5rNhmzZq_yj5-nzgKzwYjsiM47atX2TqH0mNkLGC2c6LCVCLZBWkuiJdSlzvhEI4G4bfMK3Kno2tRlhfqOVMaPGL-ZqwlVDEwqvbQH-b1EPfSqKxd3dZMim9_dDAj5ELUDU5QTF3Bs0wEUK1UTVYBZ3focWxLzdOEHtHq1EcEQEYv73n3__l91kafTAlGScabVnVrXQM5twVHeNv8SzqGuGUYNpDK8XkECdLbpV3ammRVPJgyO-qt8sdjwwgkIOjSrrV58wtbEpzDmJggVQznY8KwjbysYuJrS6ahyoKP4jMIDwslf3gSjhmpzC_iFV5VdNd9QvXnSr_n5ffVa-HEfoU-3OS3Y1Kg69PrbOcr0RLvaHngQtfwMQ8uiYvakXmPM9j1P13oqt6avcxwRSP3z88I4Uf9V4etDT7gn1I6G71KRIbV8uh8ddZFzJt5PR1c7X9WOf6gkG0Mc9KN2c-8OLNFbTp-TbrHn4bFR7cjCrOKagff1ofXT08uil68DbLEMNoZGtUsv2OkWoJFuV7AZ19u2nW7YqLKIB8KkM90o_sL3IsiAIgBGNK6FTnqFV99Xq8wA3B0IRvRJQS2O5jNwBqmIbuPkc3Pi5F9_wLCN7ax5aJ3X34ixOI4snMhc9hFPYWzqJvENcIDgI14V0hRqEIW2HYUsGM6g7J9wgAXU_fxw6zoGbyd2TnX31ZF2ddKykdcK-42hRrhlxY4otcK8mOu5SlWs0HmpEq07F8zzIS9QunQIl-8ZPEHRKptaO5FjRqkgzS_y1pwbQu7NqkX1AY73shZiXqQzD4In2npe9lrACFkMn9KUZ0oghMpTQVw5JbMMPNOxnz6Z5agS7RdTeaa10HwrogT1RIF24n53tdQqQy-4a4zMD8cc9M3TEK04Fp3ZCo_s2YB6miaehuPswdI5qxuq4E4sT9k2qDcKnRVRomJtKpCnwZw-iNHhiJLalRzoVQHPBnionaa2VZg9DtqvJg9Q1m64eWp737gH3JVC2zu1azwIA24eop-cBcHBcWCfsNhFyWDFhyBPbsuenbc3KgEaV7QxJ5UoyEzuBZuCEYIGi4YWpNefLZtaUZG1N2W62LFVieC6_LXBL68LvI5Ol5tci4447plAt-5CglVNajfaUToQ1kRrWr-tmy22UpHIDd-XGJqaDXGXcnUeuyiXAPXBy6BsB8SXNjzOschCvBV14KnFESwZmgu_c5SHj2rAxklmOXN_xX8lMFPcoNZTSJ-xHwfacouPKNC62OWodCg1qkpqjeIyyAoEsYKxFiYssRldF-0ylP_iJ7P8poGzQEKCEkW98cyX2lQC4THBTQk1JO7HyaUDKabyzvIxhQBqx_paXoM9xe6OJxgpEZINSgMqRS5pGp01UsUnQxKhQ0Z3-OgjYD2jkgn0u4N6WywLVUjLcI4XG7ZKtFBRRlOGkdKotfyA5-If5VFcMF0JWdSsjcImIGfUvlxo-xV0V0vGaqhP96oT989PXb2_fXeOIEoOyvPEGhFqgEVYAhsg-DCjuIWT7JI0STJ4C33OtVmgzZQMNpUGBFFIANEI8JYYc2g-COEXeN0PcvyKGFLN-rcWFaXGL6z4Sjjhmr3Hg3JcHCLqZ6NT4njQDDXn5fk3Q-pbtDHRIuDdOcpOGADYhwndMaucMF3nyDfGsWz1YjsckRoUR2RpNmdCCGOSKFCKmuIOcdBfIasxwAEzXhoBMp1FBD6Gx9Pba7ve9QzBerDm-LGqTCAwS9hGJTW0U3gDrBhYgDVjY6C_Oa_okmQziyZDONl1jqu3AWddFplbm_Xf-tnDXnrZw4VYHfFy-AqTY5Zr5NZu6ykm3LziOfyQ30EN9GVdgKVIR96JXXOyIYX309qBJ2REdDR1BwOM41PVFRegBfERu14lKdYdbE2qKDMnREAuHFIFRBR4pZfH6g8oSOfvHDx8_jJ0fcJF-osH9GR-Rvx7ogl4VD5LAl743NRdRuqaCElEiU_wuQfHPYb_tO1MJ3r5Lgkzx2LivlBq4K6eY2dIuZjS3ltIlLH7aoHDR21IFzSDQhewPesOeFj8XAM9dF_DrVGJtbq7H47wMIr4W400y5tqma7xDjbfRAxVNOXY30XH1WjWGJDOeXp5N5tP5xeXp6ThPc08ymgTT4GwexGI3GeXl6QjNWIy4LIN9gmmYl-QLbKAahG7Hg8e8RK-T2MhLFWMGjHaTYHYRTHHk32F710-9ql33rmfwQSn7vdrTdiUKL5ugxjDKYi1AdNec4GqNboFD97beu56eT4Y9vHb86B9P_dN3gl5s_GMag43ueX-HVCT8jZIWlzOhX9FApKOVpwVJjhBjt_eM3yIamtaKStHVPJhMrs4vrk7n86urydnpy7LB14r0V9Deodj2uBvZwD7aQws8QrX66WQWOLWTixl-27o6n7-q_9Dui3lwPj-9uLqYnp1PZ2fz6e8BswLnt4CsMTyw-xcWTE9nwdns4nRyenl-djk_n8H0pDo7vzz7Pfag2LHNs98yqKYji-jveeiCjbq5u3_-P63Ngc7hFQAA)

**Comment by antonymilne** [comment]
Newer version of the flow: https://py.cafe/antonymilne/vizro-python-to-yaml.

**Comment by huong-li-nguyen** [comment]
Shall we put this on HF or our Vizro pyCafe instance? https://py.cafe/antonymilne/vizro-python-to-yaml

**Comment by antonymilne** [comment]
Errr possibly! I was going to ask the team whether they had used it and what they thought about it/how reliable it is.

Edit: people seemed to think it worked well, maybe apart from the yaml indenting, although Cursor can also do the job well now.
